### ISSUE TITLE
Wrong Result When Adding `WITH *`

### ISSUE THREAD

**Issue by CelineWuest** [neo4j/neo4j#13137]
I found a discrepancy when running two semantically equivalent queries against an empty database:

---
1:
```cypher
UNWIND [0] AS n0 OPTIONAL MATCH (x) MERGE n1 = () RETURN 0 ORDER BY x LIMIT 0
-
RETURN COUNT { () } AS n
```
Produces:
```
╒═══╕
│"n"│
╞═══╡
│1  │
└───┘
```

---
2:
```cypher
UNWIND [0] AS n0 OPTIONAL MATCH (x) MERGE n1 = () WITH * RETURN 0 ORDER BY x LIMIT 0
-
RETURN COUNT { () }
```
Produces:
```
╒═══╕
│"n"│
╞═══╡
│0  │
└───┘
```
---

I believe the queries mentioned above are semantically equivalent and thus their result should be the same.
The only difference between the two is the inclusion of the `WITH *` clause which shouldn't have any effect on its behavior.

I encountered this issue when testing queries against the **neo4j:5.6.0** Docker container.

### Steps to reproduce
Run the aforementioned queries and observe that their output differs.

### Expected behavior
The queries should result in the same behavior

### Actual behavior
The first query returns `1`, whereas the second query returns `0`.

**Comment by AzuObs** [comment]
Hi Dominic,

Thanks a lot for your submission. I have forwarded to the right team ([private](https://trello.com/c/o4Pp6331)).

All the best,

**Comment by inmost-light** [comment]
@DominicWuest 
Thank you for reporting. This appears to be the same issue as #13151 and #13130, which should be fixed in 5.9.
